Add support for fading out document markers

pxlcoder opened the pull request:
#### 5c3f5c350c623143a75e7b21c99695aeac53d2eb
<pre>
Add support for fading out document markers
<a href="https://bugs.webkit.org/show_bug.cgi?id=252870">https://bugs.webkit.org/show_bug.cgi?id=252870</a>
rdar://105857998

Reviewed by Megan Gardner.

Add the ability to fade out document markers, similar to DataDetector highlights
and PageOverlays.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::DocumentMarkerController):

Specify constants for the animation frame rate and duration. The frame rate
was selected to match other existing animations, such as fading out DataDetector
highlights and PageOverlays.

(WebCore::DocumentMarkerController::detach):
(WebCore::DocumentMarkerController::forEachOfTypes):

Add a helper method to iterate over all markers of a given type.

(WebCore::DocumentMarkerController::removeMarkers):
(WebCore::DocumentMarkerController::removeMarkersFromList):

Introduce a filter function, to avoid code duplication when it is time to
remove faded out markers.

(WebCore::DocumentMarkerController::dismissMarkers):

This method removes markers of a given type by fading them out, rather than
simply removing them from the list of rendered document markers.

Markers that are fading out, as marked as &quot;being dismissed&quot; and an animation
start time is recorded. If the animation timer is not already running, it is
started.

(WebCore::DocumentMarkerController::fadeAnimationTimerFired):

On each tick, update the opacity of markers being dismissed based on the
animation start time and the current time. Once the marker is fully faded out,
it is removed from the list of markers. If there are no more markers to animate,
the timer is stopped.

* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebCore/dom/RenderedDocumentMarker.h:

Add new properties to `RenderedDocumentMarker` to support animation.

(WebCore::RenderedDocumentMarker::opacity const):
(WebCore::RenderedDocumentMarker::setOpacity):
(WebCore::RenderedDocumentMarker::isBeingDismissed const):
(WebCore::RenderedDocumentMarker::setBeingDismissed):
(WebCore::RenderedDocumentMarker::animationStartTime const):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintPlatformDocumentMarker):

Adjust the alpha value of the marker color to perform the fade.

Canonical link: <a href="https://commits.webkit.org/260842@main">https://commits.webkit.org/260842@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/578d25b95feab7f77a6ec9215104086325dbb82a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109642 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18762 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42387 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1123 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118765 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20229 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9961 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101911 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115395 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15057 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98292 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43273 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97031 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85053 "Found 1 new API test failure: /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11493 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31286 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12149 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8223 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17515 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50895 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7517 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13893 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

megangardner commented on Source/WebCore/dom/DocumentMarkerController.cpp:
```diff
@@ -461,6 +466,22 @@ Vector<RenderedDocumentMarker*> DocumentMarkerController::markersFor(Node& node,
     return result;
 }
 
+void DocumentMarkerController::forEach(OptionSet<DocumentMarker::MarkerType> types, const Function<bool(Node&, RenderedDocumentMarker&)> function)
```
I think this should be more like forEachofType or something else that indicates that it's not jut for all the markers. Or maybe I'm thinking too Obj-C style names here.

pxlcoder replied:
Sounds good – will rename to `forEachOfTypes` (since multiple types could be passed in the set).

megangardner commented on Source/WebCore/dom/RenderedDocumentMarker.h:
```diff
@@ -68,9 +70,32 @@ class RenderedDocumentMarker : public DocumentMarker {
+    float opacity() const { return m_opacity; }
+    void setOpacity(float opacity) { m_opacity = opacity; }
+
+    bool isActive() const { return m_isActive; }
```
I'm not sure I like 'active' for this state, especially if it's more of a transitional state that will only be around when fading. I think that 'fading/isFading' or 'deactivating/isDeactivating' is a better term for an actively transitionary state, rather than 'inactive' which makes me thing of a marker that is still around, but not being painted for some reason.

pxlcoder replied:
I will rename to `isBeingDismissed`, since that aligns with the method name `dismissMarkers`.

pxlcoder committed:
Add support for fading out document markers
https://bugs.webkit.org/show_bug.cgi?id=252870
rdar://105857998

Reviewed by Megan Gardner.

Add the ability to fade out document markers, similar to DataDetector highlights
and PageOverlays.

* Source/WebCore/dom/DocumentMarkerController.cpp:
(WebCore::DocumentMarkerController::DocumentMarkerController):

Specify constants for the animation frame rate and duration. The frame rate
was selected to match other existing animations, such as fading out DataDetector
highlights and PageOverlays.

(WebCore::DocumentMarkerController::detach):
(WebCore::DocumentMarkerController::forEachOfTypes):

Add a helper method to iterate over all markers of a given type.

(WebCore::DocumentMarkerController::removeMarkers):
(WebCore::DocumentMarkerController::removeMarkersFromList):

Introduce a filter function, to avoid code duplication when it is time to
remove faded out markers.

(WebCore::DocumentMarkerController::dismissMarkers):

This method removes markers of a given type by fading them out, rather than
simply removing them from the list of rendered document markers.

Markers that are fading out, as marked as "being dismissed" and an animation
start time is recorded. If the animation timer is not already running, it is
started.

(WebCore::DocumentMarkerController::fadeAnimationTimerFired):

On each tick, update the opacity of markers being dismissed based on the
animation start time and the current time. Once the marker is fully faded out,
it is removed from the list of markers. If there are no more markers to animate,
the timer is stopped.

* Source/WebCore/dom/DocumentMarkerController.h:
* Source/WebCore/dom/RenderedDocumentMarker.h:

Add new properties to `RenderedDocumentMarker` to support animation.

(WebCore::RenderedDocumentMarker::opacity const):
(WebCore::RenderedDocumentMarker::setOpacity):
(WebCore::RenderedDocumentMarker::isBeingDismissed const):
(WebCore::RenderedDocumentMarker::setBeingDismissed):
(WebCore::RenderedDocumentMarker::animationStartTime const):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintPlatformDocumentMarker):

Adjust the alpha value of the marker color to perform the fade.

Canonical link: https://commits.webkit.org/260842@main
